random: Updated documentation

Jjp137 opened the pull request:
Here's the updated documentation for the `random` module. There are still a few things that I'd like some feedback on, which are described in the notes below.

Notes about this pull request:
- It's not fully clear to me how to use the `skipRandomNumbers` proc effectively. The current version of the docs for this proc simply copies the corresponding comment in the C implementation, and I left it largely untouched here. The example I wrote isn't really insightful, either. Any suggestions for explaining that proc in a better way are welcome.
- I opted not to use `doAssert` for the runnable examples in this module since code such as `let x = rand(1..6); doAssert x >= 1 and x <= 6` isn't enough to convince me that the proc always works, especially since the default Rand state always starts at the same state. I'm not sure if putting every example in a loop that lasts for many iterations would be a good idea, either, since that could be annoying to wait for when updating the documentation. There are tests located at the bottom of the module already, so I just let the examples illustrate how to use each proc, although I suppose the examples help ensure that the procs don't cause crashes, at least.
- Since `doAssert` isn't being used in the runnable examples at the moment, I ended up assigning the return values to variables that end up unused, leading to a lot of `XDeclaredButNotUsed` hints. I didn't want to use `echo` or `discard` since the former would cause something to print out when `nim doc` is run and the latter seemed like noise to someone reading the docs. Can I ignore those hints, and if not, should I use `echo`, `discard`, or some other alternative?
- I get a lot of `RedefinitionOfLabel` warnings when running `nim doc` with my changes. I figured out that it is because I use the same link text several times within a proc's "See also" section, but I don't know how to address those warnings or if they should be addressed at all.
- Should I add any more modules to the top-level "See also" section?

Feedback and nitpicks are welcome :)

Refs: #10330, CC: @narimiran

narimiran commented on lib/pure/random.nim:
```diff
@@ -196,37 +391,117 @@ proc sample*[T, U](r: var Rand; a: openArray[T], cdf: openArray[U]): T =
+  ## See also:
+  ## * `randomize proc<#randomize,int64>`_ that accepts a seed for the default
+  ##   random number generator
+  ## * `randomize proc<#randomize,>`_ that initializes the default random
```
These trailing commas in the links have been fixed in the last couple of days.
Use: `randomize proc<#randomize>`

narimiran commented on lib/pure/random.nim:
```diff
@@ -7,12 +7,71 @@
+##   import random
+##
+##   # Initializes the default random number generator
+##   # Without a call to randomize(), the same results will occur every time
```
will -> would ?

Jjp137 replied:
I think "will" is alright here since, from what I read, it seems to imply more certainty than "would", and it's certain what happens in this case.

Jjp137 commented on lib/pure/random.nim:
```diff
@@ -143,51 +245,144 @@ proc rand*(r: var Rand; max: range[0.0 .. high(float)]): float {.benign.} =
-  ## Returns a random element from the openarray `a`.
-  ## **Deprecated since v0.20.0:** use ``sample`` instead.
+  ## **Deprecated since version 0.20.0:**
+  ## Use `sample[T](Rand, openArray)<#sample,Rand,openArray[T]>`_ instead.
```
Fixed in https://github.com/nim-lang/Nim/pull/10546/commits/c8346a31325716b1186b0fb25a93e2416e51298a

Jjp137 commented on lib/pure/random.nim:
```diff
@@ -7,12 +7,71 @@
+## use the default generator. Those procs are **not** thread-safe.
+##
+## Note that the default generator always starts in the same state.
+## The `randomize proc<#randomize,>`_ can be called to initialize the default
```
Good point; it became so obvious to me that I neglected to mention that.

Fixed in https://github.com/nim-lang/Nim/pull/10546/commits/c8346a31325716b1186b0fb25a93e2416e51298a

Jjp137 commented on lib/pure/random.nim:
```diff
@@ -7,12 +7,71 @@
+## cryptographic applications.
+##
+## See also
+## ========
```
Right now, I decided to just link to the section in general (in https://github.com/nim-lang/Nim/pull/10546/commits/c8346a31325716b1186b0fb25a93e2416e51298a).

Jjp137 commented:
@timotheecour

> IMO the best in most cases is to use a fixed seed and doAssert on the result directly:

I'm not sure if the `runnableExamples` should depend so tightly on the exact implementation being used though.

>     var r = initRand(5) # 5 is bad because it confuses the user as it's equal to cards.len

Changed in https://github.com/nim-lang/Nim/pull/10546/commits/c8346a31325716b1186b0fb25a93e2416e51298a; thanks

narimiran commented:
> I'm not sure if the `runnableExamples` should depend so tightly on the exact implementation being used though.

Yeah, it is not perfect, but I think it is better than not having anything. (And in the mean time I didn't manage to come up with anything better)

Since these examples are run when the documentation is created, if the implementation changes and the examples become "invalid", this will be caught early.

For some examples, to show the randomness, more than one call to a function might be a good idea, something like:
```nim
randomize(2019)
assert rand(1..6) == 4
assert rand(1..6) == 5
assert rand(1..6) == 1
```

demotomohiro commented:
State of xoroshiro128+ must not be 0 or it generates 0 forever.
I think `initRand` needs a comment like: "`seed` must not be 0".

I tried to add a this sample code under `runnableExamples` of `skipRandomNumbers` in module `random` before.
https://github.com/nim-lang/Nim/issues/9359
But it is too long for `runnableExamples` and it failed to `koch doc` because it was compiled without `--threads:on` option while it uses `spawn`.

`skipRandomNumbers` proc is used when you use random numbers in multiple threads/processes.
Each threads have to own their `Rand` object so that they can get random number safely and efficienty.
If each threads initialize `Rand` with some random numbers(e.g, time, thread/process id, memory address, /dev/random, etc), subsequences of random number could be overlap even if each seeds are unique value.

For example, suppose there are threads A and B, and each threads generate 10 random numbers.
a0, a1, ..., a8, a9 are random numbers generated in thread A and b0, b1, ..., b8, b9 are in thread B.
They initialize `Rand` with random numbers, and the state of B's `Rand` after generating 5 random numbers could be the same state of A's `Rand` before A generating first random number.
So, random number subsequence of thread A (a0, a1, a2, a3, a4) is same to the subsequence of thread B (b5, b6, b7, b8, b9).

If you generate large number of random numbers on many threads, such overlapping can happen more often.
I think the probability of getting overlapping sequence can be calculated with this if all threads generate same amount of random number:
https://en.wikipedia.org/wiki/Birthday_problem#Near_matches
(n: number of threads, k: number of random number per thread, m: the period of PRNG)

If you initialize `Rand` once and call `skipRandomNumbers` everytime passing `Rand` to each thread, such overlapping never happen when each thread generates less than 2^64 random numbers and a number of threads is less than 2^64.
Because no thread can have a state of `Rand` that other threads can have while they are using it.

Jjp137 commented:
Alright, I updated the examples and a few other things in https://github.com/nim-lang/Nim/pull/10546/commits/3acc3d4781477dd50a486adc7574d0dcf0b6e50f. The only proc I have doubts about is `rand(float)` since I'm not sure if `==` is okay in that case, so I just put what the value ended up being in a comment.

@demotomohiro:
 
> State of xoroshiro128+ must not be 0 or it generates 0 forever.
> I think `initRand` needs a comment like: "`seed` must not be 0".

Thanks for pointing that out; added.

Also, thanks for the explanation about `skipRandomNumbers`. Please check if what I wrote for that proc in https://github.com/nim-lang/Nim/pull/10546/commits/3acc3d4781477dd50a486adc7574d0dcf0b6e50f is accurate. In addition, I took a look at #9359 and added a rather simple example based on the structure of your code there. Right now, I am not going to figure out how to compile that particular example with `--threads:on` (maybe that should be addressed in a separate PR at some point?), especially since what you tried within that issue didn't seem to work, so I had to cheat a bit and place the example within the documentation comments instead.

timotheecour commented:
>> State of xoroshiro128+ must not be 0 or it generates 0 forever.
>> I think `initRand` needs a comment like: "`seed` must not be 0".

> Thanks for pointing that out; added.

I ran into this gotcha before; see https://github.com/nim-lang/Nim/pull/10635 for a follow-up that adds a `doAssert`
